### PR TITLE
Add ability to choose other models

### PR DESCRIPTION
**Pull request by ccurme**
https://github.com/langchain-ai/langchain-extract/issues/37

Here we add an argument to ExtractionRequest to specify the underlying chat model.

Alternatively we could add this as a property of an extractor, but I kept them separate so that users could easily experiment with different models for the same underlying schema.

### CODE REVIEW

File: backend/server/api/chat_models.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+router = APIRouter(
+    prefix="/chat_models",
+    tags=["Available chat models"],
+    responses={404: {"description": "Not found"}},
+)
+
+
+@router.get("")
+def list_chat_models() -> List[str]:
```
Comment (eyurtsev) [inline backend/server/api/chat_models.py:16]: I suspect we'll want to generalize this a bit. There might be other configuration that we'll want to fetch (e.g., does the server support retrieving content by URL)

File: backend/server/settings.py [hunk 2]
```diff
@@ -23,6 +20,45 @@ def get_postgres_url() -> URL:
     return url
 
 
-def get_model() -> ChatOpenAI:
+class ChatModel(NamedTuple):
```
Comment (eyurtsev) [inline backend/server/settings.py:23]: Have you considered using runnable configurable here?

File: backend/server/api/chat_models.py (hunk 1 above)
Comment (eyurtsev) [reply backend/server/api/chat_models.py:16]: We can do an end point like this:

```
def get_configurables():
   return {
     'models': []

   }
```

File: backend/server/settings.py (hunk 2 above)
Comment (ccurme) [reply backend/server/settings.py:23]: One issue I'm getting with this is if I set model to be `ChatOpenAI(model="gpt-3.5-turbo", temperature=0).configurable_alternatives(...)`, then it is now a `RunnableConfigurableAlternatives` which does not support `with_structured_output`. lmk if you had something else in mind for runnable configurable or if there's a workaround.

File: backend/server/api/chat_models.py (hunk 1 above)
Comment (ccurme) [reply backend/server/api/chat_models.py]: thanks, makes sense

File: backend/server/settings.py [hunk 3]
```diff
@@ -23,6 +20,45 @@ def get_postgres_url() -> URL:
...
+    name: str
+    chunk_size: int
+    constructor: Callable
+
+
+SUPPORTED_MODELS = (
```
[the remark this answers is not in the document]
Comment (ccurme) [reply backend/server/settings.py:30]: Thanks, added fireworks and together-ai. langchain_mistralai requires python 3.9+, how do you feel about dropping support for 3.8 here in order to support mistral?

File: backend/server/settings.py (hunk 2 above)
Comment (ccurme) [reply backend/server/settings.py:24]: Added an issue to capture this. Resolving this comment for now.

https://github.com/langchain-ai/langchain/issues/19279

File: backend/server/api/configurables.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    tags=["Available chat models"],
+    responses={404: {"description": "Not found"}},
+)
+
+
+@router.get("")
+def get_configurables() -> Dict[str, Any]:
+    """Endpoint to get all chat models."""
+    return {
+        "models": list(SUPPORTED_MODELS.keys()),
```
Comment (eyurtsev) [inline backend/server/api/configurables.py:19]:
```suggestion
        "models": sorted(SUPPORTED_MODELS.keys()),
```

File: backend/server/api/extract.py [hunk 5]
```diff
@@ -24,6 +25,7 @@ async def extract_using_existing_extractor(
     text: Optional[str] = Form(None),
     mode: Literal["entire_document", "retrieval"] = Form("entire_document"),
     file: Optional[UploadFile] = File(None),
+    model_name: Optional[ModelNameLiteral] = Form("gpt-3.5-turbo"),
```
Comment (eyurtsev) [inline backend/server/api/extract.py:28]:
```suggestion
    model_name: Optional[ModelNameLiteral] = Form("default") # <-- could we do this?
```

File: backend/server/models.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
+import os
+from typing import Literal, Optional
+
+from langchain_core.language_models.chat_models import BaseChatModel
+from langchain_fireworks import ChatFireworks
+from langchain_openai import ChatOpenAI
+
+SUPPORTED_MODELS = {
```
Comment (eyurtsev) [inline backend/server/models.py:8]: Could we create this dynamically based on which keys are present in the enviornment?

File: backend/server/models.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+        api_key=os.environ["TOGETHER_API_KEY"],
+        model="mistralai/Mixtral-8x7B-Instruct-v0.1",
+    ),
+}
+
+CHUNK_SIZES = {  # in tokens, defaults to int(4_096 * 0.8). Override here.
+    "gpt-4-0125-preview": int(128_000 * 0.8),
+}
+
+ModelNameLiteral = Literal[
```
Comment (eyurtsev) [inline backend/server/models.py:23]: This will need to be dynamic or we can just use `str`

File: backend/server/models.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+SUPPORTED_MODELS = {
+    "gpt-3.5-turbo": ChatOpenAI(model="gpt-3.5-turbo", temperature=0),
+    "gpt-4-0125-preview": ChatOpenAI(model="gpt-4-0125-preview", temperature=0),
+    "fireworks": ChatFireworks(model="accounts/fireworks/models/firefunction-v1"),
+    "together-ai-mistral-8x7b-instruct-v0.1": ChatOpenAI(
+        base_url="https://api.together.xyz/v1",
+        api_key=os.environ["TOGETHER_API_KEY"],
+        model="mistralai/Mixtral-8x7B-Instruct-v0.1",
+    ),
+}
```
Comment (eyurtsev) [inline backend/server/models.py:17]:
```suggestion

def get_supported_models(): 
   models = {}
   if 'OPENAI_API_KEY' in env: 
      models['gpt-3.5-turbo'] = ChatOpenAI()..
    if ... in os.environ:
      models[...] = ChatFireworks()
      
   return models
      
    
      
      

  

SUPPORTED_MODELS = get_supported_models()
DEFAULT_MODEL = 'gpt-3.5-turbo' # Can  hard code for now to keep simple

```

File: backend/server/settings.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ccurme) [reply backend/server/settings.py]: mistralai / langchain_mistralai needs `pydantic = "^2.5.2"`. will keep it out for now.

**Pull request merged**